std/fmt.zig

```diff
@@ -256,126 +397,111 @@ pub fn formatType(
-                if (@typeOf(int_value).bit_count <= 8) {
-                    if (fmt.len > 1)
-                        @compileError("Unknown format character: " ++ [_]u8{fmt[1]});
-                    return formatAsciiChar(u8(int_value), context, Errors, output);
-                }
-            },
-            'b' => {
-                radix = 2;
-                uppercase = false;
-                width = 0;
-            },
-            'd' => {
-                radix = 10;
-                uppercase = false;
-                width = 0;
-            },
-            'x' => {
-                radix = 16;
-                uppercase = false;
-                width = 0;
-            },
-            'X' => {
-                radix = 16;
-                uppercase = true;
-                width = 0;
-            },
-            else => @compileError("Unknown format character: " ++ [_]u8{fmt[0]}),
+    if (fmt.len == 0 or comptime std.mem.eql(u8, fmt, "d")) {
+        radix = 10;
+        uppercase = false;
+    } else if (comptime std.mem.eql(u8, fmt, "c")) {
+        if (@typeOf(int_value).bit_count <= 8) {
+            return formatAsciiChar(u8(int_value), context, Errors, output);
+        } else {
+            @compileError("Cannot print integer that is larger than 8 bits as a ascii");
         }
-        if (fmt.len > 1) width = comptime (parseUnsigned(usize, fmt[1..], 10) catch unreachable);
+    } else if (comptime std.mem.eql(u8, fmt, "b")) {
+        radix = 2;
+        uppercase = false;
+    } else if (comptime std.mem.eql(u8, fmt, "x")) {
+        radix = 16;
+        uppercase = false;
+    } else if (comptime std.mem.eql(u8, fmt, "X")) {
+        radix = 16;
+        uppercase = true;
+    } else {
+        @compileError("Unknown format string: '" ++ fmt ++ "'");
     }
-    return formatInt(int_value, radix, uppercase, width, context, Errors, output);
+
+    if (options.width) |w| return formatInt(int_value, radix, uppercase, w, context, Errors, output);
+    return formatInt(int_value, radix, uppercase, 0, context, Errors, output);
 }
 
 fn formatFloatValue(
     value: var,
     comptime fmt: []const u8,
+    comptime options: FormatOptions,
     context: var,
     comptime Errors: type,
     output: fn (@typeOf(context), []const u8) Errors!void,
 ) Errors!void {
-    comptime var width: ?usize = null;
-    comptime var float_fmt = 'e';
-    if (fmt.len > 0) {
-        float_fmt = fmt[0];
-        if (fmt.len > 1) width = comptime (parseUnsigned(usize, fmt[1..], 10) catch unreachable);
-    }
-
-    switch (float_fmt) {
-        'e' => try formatFloatScientific(value, width, context, Errors, output),
-        '.' => try formatFloatDecimal(value, width, context, Errors, output),
-        else => @compileError("Unknown format character: " ++ [_]u8{float_fmt}),
+    if (fmt.len == 0 or comptime std.mem.eql(u8, fmt, "e")) {
+        // TODO: bug: Accessing options.precision directly results in null even with a visible payload.
```

Comment:
Fairly sure this is a bug. When I log the value of `options.precision` via a warn I get null as the result even if it has a non-null payload. I think this was also the case with a `@compileLog` but would need to double-check. Regardless, I would expect to be able to pass `options.precision` directly with no issue.